docs/docs/adrs/adr-008-throttle-retries.md

```diff
@@ -0,0 +1,98 @@
+---
+sidebar_position: 7
+title: Throttle with retries
+---
+
+# ADR {008}: {Throttle with retries}
+
+## Changelog
+
+* {6/9/23}: Initial draft
+
+## Status
+
+> A decision may be "proposed" if it hasn't been agreed upon yet, or "accepted" once it is agreed upon. If a later ADR changes or reverses a decision, it may be marked as "deprecated" or "superseded" with a reference to its replacement.
+
+Proposed
+
+## Context
+
+For context on why the throttling mechanism exists, see [ADR 002](./adr-002-throttle.md).
+
+Note the terms slash throttling and jail throttling are synonymous, since in replicated security a `SlashPacket` simply jails a validator for downtime infractions.  
+
+Currently the throttling mechanism is designed so that provider logic (slash meter, etc.) dictates how many slash packets can be handled over time. Throttled slash packets are persisted on the provider, leading to multiple possible issues. Namely:
+
+* If slash or vsc matured packets are actually throttled/queued on the provider, state can grow and potentially lead to a DoS attack. We have short term solutions around this, but overall they come with their own weaknesses. See [#594](https://github.com/cosmos/interchain-security/issues/594).
+* If a jailing attack described in [ADR 002](adr-002-throttle.md) were actually to be carried out with the current throttling design, we'd likely have to halt the provider, and perform an emergency upgrade and/or migration to clear the queues of slash packets that were deemed to be malicious. Alternatively, validators would just have to _tough it out_ and wait for the queues to clear, during which all/most validators would be jailed. Right after being jailed, vals would have to unjail themselves promptly to ensure safety. The synchronous coordination required to maintain safety in such a scenario is not ideal.
+
+So what's the solution? We can improve the throttling mechanism to instead queue/persist relevant data on each consumer, and have consumers retry slash requests as needed.
+
+## Decision
+
+> This section explains all of the details of the proposed solution, including implementation details.
+It should also describe affects / corollary items that may need to be changed as a part of this.
+If the proposed change will be large, please also indicate a way to do the change to maximize ease of review.
+(e.g. the optimal split of things to do between separate PR's)
+
+### Provider changes
+
+The main change needed for the provider is the removal of queuing logic for slash and vsc matured packets upon being received. Instead, the provider will consult the slash meter to determine if a slash packet can be handled immediately. If not, the provider will return an ack message to the consumer communicating that the slash packet could not be handled, and needs to be sent again in the future (retried). VSCMatured packets will always be handled immediately upon being received by the provider.
```

Comment:
> VSCMatured packets will always be handled immediately upon being received by the provider.

This is not true. Only if they are accepted. Once the slash meter is negative, all packets will be dropped until the meter is replenished. This is necessary as the provider cannot trust the consumer.

Reply:
No, I would not drop VSCMatured packets. That whole mechanism made the existing throttling code 5x more complicated. With cryptographic verification of equivocation, this whole mechanism will exist only to handle jailings, and I don't think we need to guarantee the unbonding period for those.

Reply:
Here's an explanation of why the provider should never drop VSCMatured packets, and always handle them immediately, in the context of this ADR proposal:

First we answer, what does a VSCMatured packet communicate to the provider? A VSCMatured packet communicates that a VSC has been applied to a consumer long enough that infractions committed on the consumer could have been submitted.

## If the consumer is following the queuing/blocking protocol desribed

No bad behavior happens, `VSC Maturity and Slashing Order` property is maintained. 

## If a consumer sends VSCMatured packets too leniently

The consumer is malicious and sending duplicate vsc matured packets, or sending them sooner than they should be. In this scenario, the provider needs to handle vsc matured packets immediately to prevent DOS, state bloat, or other issues. The only possible negative outcome is that the malicious consumer may not be able to jail a validator who should have been jailed. The malicious behavior only creates a negative outcome for the chain that is being malicious. Seems reasonable to me.

## If a consumer blocks the sending of VSCMatured packets

The consumer is malicious and blocking vsc matured packets that should have been sent. This will block unbonding only up until the vsc timeout is triggered. At that time the consumer is removed. Again the malicious behavior only creates a negative outcome for the chain that is being malicious

Reply:
Thanks @smarshall-spitzbart. This seems reasonable and I agree we should go with your proposed solution. Maybe add this reasoning to the ADR.

Reply:
I see that you already done that. Great :)